### PR TITLE
[RFC] Reduce overhead of the mutations for SELECTs (v2)

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reduce overhead of the mutations for SELECTs (v2)

SELECTs are affected by the mutations, since it tries to apply them on fly, and scanning over existing mutations can take significant amount of time (for simple queries, i.e. count())

And also even after mutation had been finished, it still a problem, because mutations are not removed instantly.

So instead introduce an atomic counter alter_conversions_mutations, that is incremented for new mutations and decremented once mutation finished/killed, that way once the mutation finished they will not affect queries.

Here are some numbers for non-RENAME mutations:

    rmt vanilla w/o mutations | queries: 3693, QPS: 494.813
    rmt vanilla w/ mutations  | queries: 2190, QPS: 388.256
    rmt patched w/o mutations | queries: 3168, QPS: 620.061
    rmt patched w/ mutations  | queries: 3155, QPS: 614.424
    mt vanilla w/o mutations  | queries: 3498, QPS: 656.399
    mt vanilla w/ mutations   | queries: 3821, QPS: 600.425
    mt patched w/o mutations  | queries: 5732, QPS: 745.585
    mt patched w/ mutations   | queries: 4719, QPS: 715.034

Follow-up for: https://github.com/ClickHouse/ClickHouse/pull/59531 (cc @al13n321)

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 9365cc7f1355d33fc0cdf7b3e53d8dab70d116c7 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60856/9365cc7f1355d33fc0cdf7b3e53d8dab70d116c7/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by azat** [comment]
>Performance Comparison Aarch64 [1/4] — Failed to parse the report.

```
2024-03-07 01:09:58      + clickhouse-local --version
2024-03-07 01:09:58      aarch64-binfmt-P: Could not open '/lib/ld-linux-aarch64.so.1': No such file or directory
```

>Performance Comparison [2/4] — Errors while building the report.

```
2024.03.07 02:09:53.570417 [ 33827 ] {} <Fatal> BaseDaemon: (version 24.3.1.633, build id: D084CF475099C7A68E9A9BA4788ADBC9A7A28DD4, git hash: 401df51281e581936803bcce36d3c6ada9b7210f) (from thread 34439) Terminate called without an active exception
2024.03.07 02:09:53.570615 [ 34441 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2024.03.07 02:09:53.570639 [ 34441 ] {} <Fatal> BaseDaemon: (version 24.3.1.633, build id: D084CF475099C7A68E9A9BA4788ADBC9A7A28DD4, git hash: 401df51281e581936803bcce36d3c6ada9b7210f) (from thread 34439) Received signal 6
2024.03.07 02:09:53.570654 [ 34441 ] {} <Fatal> BaseDaemon: Signal description: Aborted
2024.03.07 02:09:53.570660 [ 34441 ] {} <Fatal> BaseDaemon: 
2024.03.07 02:09:53.570674 [ 34441 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f286fc169fc 0x00007f286fbc2476 0x00007f286fba87f3 0x000000000cf7500e 0x00000000174a30e3 0x00000000174a30db 0x0000000012ddb1a3 0x0000000012ddaccd 0x0000000012ddc74a 0x0000000014eef34a 0x0000000014eeea00 0x0000000012d72c3b 0x0000000014ee91da 0x00007f286fc14ac3 0x00007f286fca6850
2024.03.07 02:09:53.570680 [ 34441 ] {} <Fatal> BaseDaemon: ########################################
2024.03.07 02:09:53.570686 [ 34441 ] {} <Fatal> BaseDaemon: (version 24.3.1.633, build id: D084CF475099C7A68E9A9BA4788ADBC9A7A28DD4, git hash: 401df51281e581936803bcce36d3c6ada9b7210f) (from thread 34439) (no query) Received signal Aborted (6)
2024.03.07 02:09:53.570695 [ 34441 ] {} <Fatal> BaseDaemon: 
2024.03.07 02:09:53.570701 [ 34441 ] {} <Fatal> BaseDaemon: Stack trace: 0x00007f286fc169fc 0x00007f286fbc2476 0x00007f286fba87f3 0x000000000cf7500e 0x00000000174a30e3 0x00000000174a30db 0x0000000012ddb1a3 0x0000000012ddaccd 0x0000000012ddc74a 0x0000000014eef34a 0x0000000014eeea00 0x0000000012d72c3b 0x0000000014ee91da 0x00007f286fc14ac3 0x00007f286fca6850
2024.03.07 02:09:53.570712 [ 34441 ] {} <Fatal> BaseDaemon: 2. ? @ 0x00007f286fc169fc
2024.03.07 02:09:53.570717 [ 34441 ] {} <Fatal> BaseDaemon: 3. ? @ 0x00007f286fbc2476
2024.03.07 02:09:53.570723 [ 34441 ] {} <Fatal> BaseDaemon: 4. ? @ 0x00007f286fba87f3
2024.03.07 02:09:53.581281 [ 34441 ] {} <Fatal> BaseDaemon: 5.0. inlined from ./contrib/llvm-project/libcxxabi/src/cxa_exception.cpp:670: __cxa_decrement_exception_refcount
2024.03.07 02:09:53.581294 [ 34441 ] {} <Fatal> BaseDaemon: 5.1. inlined from ./contrib/llvm-project/libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:17: ~exception_ptr
2024.03.07 02:09:53.581301 [ 34441 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Daemon/BaseDaemon.cpp:591: terminate_handler() @ 0x000000000cf7500e
2024.03.07 02:09:53.582233 [ 34441 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./contrib/llvm-project/libcxxabi/src/cxa_handlers.cpp:61: std::__terminate(void (*)()) @ 0x00000000174a30e3
2024.03.07 02:09:54.787376 [ 34441 ] {} <Fatal> BaseDaemon: 7. ? @ 0x00000000174a30db
2024.03.07 02:09:54.805941 [ 34441 ] {} <Fatal> BaseDaemon: 8. ./build_docker/./src/Coordination/KeeperStateMachine.cpp:137: DB::(anonymous namespace)::assertDigest(DB::KeeperStorage::Digest const&, DB::KeeperStorage::Digest const&, Coordination::ZooKeeperRequest const&, unsigned long, bool) (.llvm.10545939782211699635) @ 0x0000000012ddb1a3
2024.03.07 02:09:54.823507 [ 34441 ] {} <Fatal> BaseDaemon: 9.0. inlined from ./contrib/llvm-project/libcxx/include/atomic:958: bool std::__cxx_atomic_load[abi:v15000]<bool>(std::__cxx_atomic_base_impl<bool> const*, std::memory_order)
2024.03.07 02:09:54.823521 [ 34441 ] {} <Fatal> BaseDaemon: 9.1. inlined from ./contrib/llvm-project/libcxx/include/atomic:1588: std::__atomic_base<bool, false>::load[abi:v15000](std::memory_order) const
2024.03.07 02:09:54.823528 [ 34441 ] {} <Fatal> BaseDaemon: 9.2. inlined from ./src/Common/ThreadFuzzer.cpp:186: DB::ThreadFuzzer::isStarted()
2024.03.07 02:09:54.823535 [ 34441 ] {} <Fatal> BaseDaemon: 9.3. inlined from ./src/Common/ThreadFuzzer.cpp:436: pthread_mutex_unlock
2024.03.07 02:09:54.823542 [ 34441 ] {} <Fatal> BaseDaemon: 9.4. inlined from ./contrib/llvm-project/libcxx/include/__threading_support:314: std::__libcpp_mutex_unlock[abi:v15000](pthread_mutex_t*)
2024.03.07 02:09:54.823549 [ 34441 ] {} <Fatal> BaseDaemon: 9.5. inlined from ./contrib/llvm-project/libcxx/src/mutex.cpp:52: std::mutex::unlock()
2024.03.07 02:09:54.823555 [ 34441 ] {} <Fatal> BaseDaemon: 9.6. inlined from ./contrib/llvm-project/libcxx/include/__mutex_base:100: ~lock_guard
2024.03.07 02:09:54.823561 [ 34441 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Coordination/KeeperStateMachine.cpp:299: DB::KeeperStateMachine::preprocess(DB::KeeperStorage::RequestForSession const&) @ 0x0000000012ddaccd
2024.03.07 02:09:54.836503 [ 34441 ] {} <Fatal> BaseDaemon: 10. ./build_docker/./src/Coordination/KeeperStateMachine.cpp:389: DB::KeeperStateMachine::commit(unsigned long, nuraft::buffer&) @ 0x0000000012ddc74a
2024.03.07 02:09:54.841419 [ 34441 ] {} <Fatal> BaseDaemon: 11.0. inlined from ./build_docker/./contrib/NuRaft/src/handle_commit.cxx:335: nuraft::raft_server::commit_app_log(unsigned long, std::shared_ptr<nuraft::log_entry>&, bool, bool)
2024.03.07 02:09:54.841430 [ 34441 ] {} <Fatal> BaseDaemon: 11. ./build_docker/./contrib/NuRaft/src/handle_commit.cxx:265: nuraft::raft_server::commit_in_bg_exec(unsigned long, bool) @ 0x0000000014eef34a
2024.03.07 02:09:54.846723 [ 34441 ] {} <Fatal> BaseDaemon: 12. ./build_docker/./contrib/NuRaft/src/handle_commit.cxx:158: nuraft::raft_server::commit_in_bg() @ 0x0000000014eeea00
2024.03.07 02:09:54.865039 [ 34441 ] {} <Fatal> BaseDaemon: 13.0. inlined from ./src/Common/LockMemoryExceptionInThread.cpp:18: ~LockMemoryExceptionInThread
2024.03.07 02:09:54.865051 [ 34441 ] {} <Fatal> BaseDaemon: 13. ./build_docker/./src/Coordination/KeeperServer.cpp:199: DB::KeeperServer::KeeperRaftServer::commit_in_bg() @ 0x0000000012d72c3b
2024.03.07 02:09:54.878588 [ 34441 ] {} <Fatal> BaseDaemon: 14.0. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>::reset[abi:v15000](std::__thread_struct*)
2024.03.07 02:09:54.878600 [ 34441 ] {} <Fatal> BaseDaemon: 14.1. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2024.03.07 02:09:54.878607 [ 34441 ] {} <Fatal> BaseDaemon: 14.2. inlined from ./contrib/llvm-project/libcxx/include/tuple:265: ~__tuple_leaf
2024.03.07 02:09:54.878613 [ 34441 ] {} <Fatal> BaseDaemon: 14.3. inlined from ./contrib/llvm-project/libcxx/include/tuple:538: ~tuple
2024.03.07 02:09:54.878623 [ 34441 ] {} <Fatal> BaseDaemon: 14.4. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>>::operator()[abi:v15000](std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>*) const
2024.03.07 02:09:54.878646 [ 34441 ] {} <Fatal> BaseDaemon: 14.5. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>, std::default_delete<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>>>::reset[abi:v15000](std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>*)
2024.03.07 02:09:54.878653 [ 34441 ] {} <Fatal> BaseDaemon: 14.6. inlined from ./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2024.03.07 02:09:54.878659 [ 34441 ] {} <Fatal> BaseDaemon: 14. ./contrib/llvm-project/libcxx/include/thread:297: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, std::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>>(void*) @ 0x0000000014ee91da
2024.03.07 02:09:54.878668 [ 34441 ] {} <Fatal> BaseDaemon: 15. ? @ 0x00007f286fc14ac3
2024.03.07 02:09:54.878673 [ 34441 ] {} <Fatal> BaseDaemon: 16. ? @ 0x00007f286fca6850
2024.03.07 02:09:54.878680 [ 34441 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2024.03.07 02:09:54.878689 [ 34441 ] {} <Information> SentryWriter: Not sending crash report
2024.03.07 02:09:54.878695 [ 34441 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2024.03.07 02:09:58.685066 [ 33821 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

File: tests/queries/0_stateless/02555_davengers_rename_chain.sh [hunk 1]
```diff
@@ -3,141 +3,101 @@
...
+        ALTER TABLE $table RENAME COLUMN a1 TO b, RENAME COLUMN b1 to a SETTINGS replication_alter_partitions_sync = 0;
+    "
+
+    wait_mutation_loaded "$table" "b1 TO a" || exit 2
+
+    $CLICKHOUSE_CLIENT -n --query="
+        -- { echoOn }
+        SELECT 'ECHO_ALIGNMENT_FIX' FORMAT Null;
+
+        INSERT INTO $table VALUES (7, 8, 9);
```
[the remark this answers is not in the document]
Comment (azat) [reply tests/queries/0_stateless/02555_davengers_rename_chain.sh:89]: I don't like the why it is fixed very much, @tavplubix maybe you have better idea? (you can find description in the linked commit message)

**Comment by canhld94** [comment]
Interesting. Internally, we also had this problem before, but our solution is fairly naive: disable applying mutation on-the-fly.
We noticed that ClickHouse already has a setting `apply_mutations_on_fly` but only implemented for ClickHouse cloud. So we implemented it ourself.

**Comment by mrodonnpaw** [comment]
> Interesting. Internally, we also had this problem before, but our solution is fairly naive: disable applying mutation on-the-fly. We noticed that ClickHouse already has a setting `apply_mutations_on_fly` but only implemented for ClickHouse cloud. So we implemented it ourself.

Hi @canhld94 ! Can you share the changes to implement this setting? Thanks!

File: tests/queries/0_stateless/02555_davengers_rename_chain.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (tavplubix) [reply tests/queries/0_stateless/02555_davengers_rename_chain.sh:89]: I don't like it too
We have `IMergeTreeDataPart::getMetadataVersion`, maybe we can use it instead?

[the remark this answers is not in the document]
Comment (azat) [reply tests/queries/0_stateless/02555_davengers_rename_chain.sh:89]: Should work, let's try.

[the remark this answers is not in the document]
Comment (azat) [reply tests/queries/0_stateless/02555_davengers_rename_chain.sh:89]: Yes, this should work now. @tavplubix maybe you can take a look at the whole PR?

File: src/Storages/MutationCommands.cpp [hunk 2]
```diff
@@ -207,6 +207,8 @@ std::shared_ptr<ASTExpressionList> MutationCommands::ast(bool with_pure_metadata
     auto res = std::make_shared<ASTExpressionList>();
     for (const MutationCommand & command : *this)
     {
+        if (!command.ast)
+            continue;
```
Comment (tavplubix) [inline src/Storages/MutationCommands.cpp:211]: When is it possible? Does it mean that some commands may be lost when serializing?

File: src/Storages/MutationCommands.cpp [hunk 3]
```diff
@@ -217,7 +219,9 @@ std::shared_ptr<ASTExpressionList> MutationCommands::ast(bool with_pure_metadata
 void MutationCommands::writeText(WriteBuffer & out, bool with_pure_metadata_commands) const
 {
     WriteBufferFromOwnString commands_buf;
-    formatAST(*ast(with_pure_metadata_commands), commands_buf, /* hilite = */ false, /* one_line = */ true);
+    const auto commands_ast = ast(with_pure_metadata_commands);
+    if (commands_ast)
```
Comment (tavplubix) [inline src/Storages/MutationCommands.cpp:223]: Can `ast` return nullptr?

**Review by tavplubix — APPROVED** [review]

File: src/Storages/MutationCommands.cpp (hunk 2 above)
Comment (azat) [reply src/Storages/MutationCommands.cpp:211]: After digging though the reflog I finally remember the reason, I was using `waitForMutation(maybe_mutation_commands)` in `StorageMergeTree::alter()` - https://github.com/azat/ClickHouse/commit/9365cc7f1355d33fc0cdf7b3e53d8dab70d116c7#diff-205c1b8d661c01d47b9831ecd03be2e25db3a49192f186a00c8896e13eac77efL381

And I cannot reproduce any crash with the current version of the patch (even with some extra debugging logging to activate `MutationsCommands::toString`), will remove this patch and rebase.

**Comment by azat** [comment]
Test failures does not looks related:

>Integration tests (asan, old analyzer) [6/6] — fail: 5, passed: 521

Only `test_backup_restore_azure_blob_storage` failed

**Comment by clickgapai** [comment]
Hi @azat @tavplubix — while reviewing this PR I found the following:

- **Coverage gap:** `active_on_fly_alter_mutations` counter for `READ_COLUMN`/`ALTER MODIFY COLUMN` type-change is untested — https://github.com/ClickHouse/ClickHouse/pull/103912

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**